src/launcher/plugin.ts

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  h, VNode
+} from 'phosphor/lib/ui/vdom';
+
+import {
+  VDomModel, VDomWidget
+} from '../common/vdom';
+
+import {
+  ObservableList
+} from '../common/observablelist';
+
+import {
+  DisposableDelegate, IDisposable
+} from 'phosphor/lib/core/disposable';
+
+import {
+  JupyterLab, JupyterLabPlugin
+} from '../application';
+
+import {
+  ICommandPalette
+} from '../commandpalette';
+
+import {
+  IPathTracker
+} from '../filebrowser';
+
+import {
+  IServiceManager
+} from '../services';
+
+/**
+ * The landing page extension.
+ */
+export
+const launcherExtension: JupyterLabPlugin<void> = {
+  id: 'jupyter.extensions.launcher',
+  requires: [IServiceManager, IPathTracker, ICommandPalette],
+  activate: activateLauncher,
+  autoStart: true
+};
+
+
+
+function activateLauncher(app: JupyterLab, services: IServiceManager, pathTracker: IPathTracker, palette: ICommandPalette): void {
+  let launcherModel = new LauncherModel(pathTracker);
+  let launcherWidget = new LauncherWidget()
+
+  launcherWidget.model = launcherModel;
+  launcherWidget.id = 'landing-jupyterlab-widget';
+  launcherWidget.title.label = 'Launcher';
+  launcherWidget.addClass('jp-Launcher');
+
+  // hardcoded defaults
+  let names = [
+      'Notebook',
+      'Console',
+      'Terminal',
+      'Text Editor',
+  ];
+
+  let actions = [
+      'file-operations:new-notebook',
+      `console:create-${services.kernelspecs.default}`,
+      'terminal:create-new',
+      'file-operations:new-text-file',
+  ]
+
+  let list : IDisposable[] = []; //DEMO
+  let l : IDisposable; //DEMO
+
+  for (let i in names) {
+      let itemName = names[i];
+      let action = actions[i];
+      l = launcherModel.add(itemName, () => app.commands.execute(action, void 0));
+      list.push(l) // DEMO
+  }
+
+  // DEMO
+  launcherModel.add("Add Random", () => {
+
+      let index = Math.floor(Math.random() * actions.length);
+      let itemName = names[index];
+      let action = actions[index];
+      l = launcherModel.add(itemName, () => app.commands.execute(action, void 0));
+      list.push(l);
+  });
+
+  launcherModel.add("Remove Last", () => {
+      list.pop().dispose();
+  });
+  //  end DEMO
+
+
+
+  app.commands.addCommand('jupyterlab-launcher:add-item', {
+    label: 'Add Launcher Item',
+    execute: args => {
+        // I'm not sure how to handle this -pi
+       // launcherModel.add(args['name'] as string, args['clickCallback']);
```

Comment:
Ah, right, arguments to the execute function must be JSON-serializable.  This means that you have to pass the id of a command to execute on click.